guide_drafts/schema-in-depth.md

```diff
@@ -0,0 +1,218 @@
+
+`infer_schema!` is a macro provided by `diesel_codegen` when you have enabled the feature for one or more database backends. The macro will establish a database connection at compile time, query for a list of all the tables, and generate `infer_table_from_schema!` for each one. `infer_schema` will skip any table names which start with `__`.
+
+`infer_table_from_schema!` has a similar role. It will establish a database connection at compile time, and query the database for the list of columns for the given table. It will then use this information to generate an invocation of `table!` for you.
+
+You can get the same `table!` invocations that `infer_schema!` would have generated by running `diesel print-schema`. If you do not want to require a database with the proper schema be present to compile your app, `diesel print-schema` is the recommended way to create your schema file. Typically these invocations would be placed in `src/schema.rs`.
+
+`table!` is where the bulk of the code gets generated. If you wanted to, you could see the actual exact code that gets generated by running `cargo rustc -- -Z unstable-options --pretty=expanded`. However, the output will be quite noisy, and there's a lot of code that won't actually be relevant to you. Instead we're going to go step by step through a *simplified* version of this output, which only has the code which you would use directly.
+
+For this example, we'll look at the code generated by this `table!` invocation:
+
+```rust
+table! {
+    users {
+        id -> Integer,
+        name -> Text,
+        hair_color -> Nullable<Text>,
+    }
+}
+```
+
+If you just want to see the full simplified code and look through it yourself, you will find it at the end of this guide.
+
+The output of `table!` is always a Rust module with the same name. If you are using `infer_schema!` there will actually be one additional level of nesting like this:
+
+```rust
+mod infer_users {
+    pub mod users {
+        /* ... */
+    }
+}
+pub use self::infer_users::users;
+```
+
+However, if you're using `table!` directly, you will just have `mod users`. The first and most important part of this module will be the definition of the table itself:
+
+```rust
+pub struct table;
+```
+
+This is the struct that represents the user's table for the purpose of constructing SQL queries. It's usually referenced in code as `users::table` (or sometimes just `users`, more on that in a bit). Next, we'll see there's a module called `columns`, with one struct per column of the table.
+
+```rust
+pub struct id;
+pub struct name;
+pub struct hair_color;
+```
+
+Each of these structs uniquely represents each column of the table for the purpose of constructing SQL queries. Each of these structs will implement a trait called [`Expression`][], which indicates the SQL type of the column.
```

Comment:
Explain where types like `Integer` and `Text` come from? (or did you mention `diesel::types` anywhere?)